File: src/main/kotlin/net/squarelabs/pgrepl/services/SnapshotService.kt
package net.squarelabs.pgrepl.services

import net.squarelabs.pgrepl.model.*
import org.postgresql.core.BaseConnection
import org.postgresql.core.ServerVersion
import org.postgresql.replication.LogSequenceNumber
import java.sql.Connection
import java.sql.SQLException
import kotlin.reflect.KClass

class SnapshotService {

    fun takeSnapshot(con: BaseConnection, includeRows: Boolean = true): Snapshot {
        val getSchema = this.javaClass.getResource("/queries/getSchema.sql").readText()
        val schema: List<SchemaRow> = query(getSchema, SchemaRow::class, con)
        val tableNames = schema
                .map { it.tableName }
                .distinct()
        val tables: List<Table> = tableNames
                .map { tableName ->
                    val columns = schema
                            .filter { it.tableName == tableName }
                            .sortedBy { it.ordinalPosition }
                            .map { Column(it.columnName, it.dataType, it.pkOrdinal) }
                    val colNames = columns.map { it.name }
                    val rows = if(includeRows) selectAll(tableName, colNames, con) else ArrayList()
                    Table(tableName, columns, rows)
                }
        val lsn = getCurrentLSN(con)
        return Snapshot(lsn, tables)
    }

    // TODO: dedupe
    @Throws(SQLException::class)
    private fun getCurrentLSN(con: BaseConnection): Long {
        val tenPlus = con.haveMinimumServerVersion(ServerVersion.v10)
        val func = if (tenPlus) "pg_current_wal_lsn()" else "pg_current_xlog_location()"
        val sql = "SELECT ${func}"
        con.createStatement().use { st ->
            st.executeQuery(sql).use { rs ->
                return if (rs.next()) {
                    val lsn = rs.getString(1)
                    LogSequenceNumber.valueOf(lsn).asLong() + 1L
                } else {
                    LogSequenceNumber.INVALID_LSN.asLong()
                }
            }
        }
    }

    fun selectAll(tableName: String, columns: List<String>, con: Connection): List<Row> {
        val colNames = columns.joinToString(",")
        val sql = "select ${colNames} from ${tableName}"
        val rows = ArrayList<Row>()
        con.createStatement().use { st ->
            st.executeQuery(sql).use {
                while (it.next()) {
                    val args = columns
                            .map { col -> it.getObject(col) }
                            .map { if(it is Int) it.toDouble() else it } // JSON will turn to double anyway
                    val row = Row(args)
                    rows.add(row)
                }
            }
        }
        return rows
    }

    fun <T> query(query: String, clazz: KClass<*>, con: Connection): ArrayList<T> {
        val constructor = clazz.constructors.first()
        val params = constructor.parameters
        val rows = ArrayList<T>()
        con.prepareStatement(query).use { stmt ->
            stmt.executeQuery().use { rs ->
                while (rs.next()) {
                    val args = params.map { param -> rs.getObject(param.name) }
                    val inst: T = constructor.call(*args.toTypedArray()) as T
                    rows.add(inst)
                }
            }
        }
        return rows
    }
}

File: src/services/SnapshotService.mjs
import fs from 'fs';

import { db } from './DbService.mjs';

const GET_SCHEMA = fs.readFileSync(`./queries/getSchema.sql`, `utf8`);

class SnapshotService {
    async takeSnapshot(primaryTableName, pk) {
        const [columns, relations] = await db.multi(GET_SCHEMA);
        const tableNames = [...new Set(columns.map(t => t.tableName))];
        const tables = tableNames.map(tableName => {
            const cols = columns.filter(col => col.tableName === tableName);
            cols.sort((a, b) => a.ordinalPosition < b.ordinalPosition);
            return { name: tableName, columns: cols, rows: [] };
        });
        const lsn = await this.getCurrentLSN();
        if (primaryTableName && pk) {
            const primaryTable = tables.find(t => t.name === primaryTableName);
            const constraintNames = [...new Set(relations.map(r => r.constraint))];
            for(const constraintName of constraintNames) {
                const fkCols = relations.filter(r => r.constraint === constraintName);
                const fkColNames = fkCols.map(c => c.foreignColumn);
                const foreignTableName = fkCols[0].foreignTable;
                const foreignTable = tables.find(t => t.name === foreignTableName);
                const colNames = foreignTable.columns.map(it => it.columnName);
                const results = await this.select(foreignTable.name, colNames, fkColNames, pk);
                foreignTable.rows.push(...results);
            }
            const pkCols = columns
                .filter(col => col.tableName === primaryTable.name && col.pkOrdinal)
                .map(col => col.columnName);
            const colNames = columns
                .filter(col => col.tableName === primaryTable.name)
                .map(col => col.columnName);
            primaryTable.rows = await this.select(primaryTable.name, colNames, pkCols, pk);
        }
        return { lsn, tables };
    }

    async select(tableName, columns, keys, values) {
        const clause = keys.map(k => `"${k}"=$1`).join(` and `);
        const colNames = columns.map(name => `"${name}"`).join(`,`);
        const sql = `select ${colNames} from "${tableName}" where ${clause}`;
        const rows = await db.any(sql, ...values);
        return rows;
    }

    async selectAll(tableName, columns) {
        const colNames = columns.map(name => `"${name}"`).join(`,`);
        const sql = `select ${colNames} from "${tableName}"`;
        const rows = await db.any(sql);
        return rows;
    }

    async getVersion() {
        if (this.version) return this.version;
        const text = (await db.one(`SELECT version();`)).version;
        const group = /PostgreSQL\s([0-9]*)/g.exec(text)[1];
        this.version = parseInt(group, 10);
        return this.version;
    }

    async getCurrentLSN() {
        const func = (await this.getVersion()) >= 10 ? `pg_current_wal_lsn` : `pg_current_xlog_location`;
        const lsn = (await db.one(`SELECT ${func}()`))[func];
        return lsn;
    }
}

export default new SnapshotService();